server/src/main/java/org/opensearch/snapshots/SnapshotsService.java

```diff
@@ -1769,10 +1770,49 @@ public ClusterState execute(ClusterState currentState) throws Exception {
                     snapshotNames,
                     repoName
                 );
-                deleteFromRepoTask = createDeleteStateUpdate(snapshotIds, repoName, repositoryData, Priority.NORMAL, listener);
+                final List<SnapshotId> snapshotIdNotBackingIndex = filterSnapshotsBackingAnyIndex(currentState, snapshotIds);
+                deleteFromRepoTask = createDeleteStateUpdate(
+                    snapshotIdNotBackingIndex,
+                    repoName,
+                    repositoryData,
+                    Priority.NORMAL,
+                    listener
+                );
                 return deleteFromRepoTask.execute(currentState);
             }
 
+            private List<SnapshotId> filterSnapshotsBackingAnyIndex(ClusterState currentState, List<SnapshotId> snapshotIds) {
```

Comment:
On a deep dive, it looks like there is no filtration going here. It is a binary operation where anytime you encounter a remote index with the same snapshot ID, an exception is thrown.

This can be simplified in terms of looping and structures as - 

```suggestion
            private void verifySnapshotsBackingAnyIndex(ClusterState currentState, List<SnapshotId> snapshotIds) {
                
                ImmutableOpenMap<String, IndexMetadata> indicesMap = currentState.getMetadata().getIndices();
                Set<String> indexSearchableSnapshotIds = new HashSet<>();
                indicesMap.valuesIt().forEachRemaining(indexMetadata -> {
                    String storeType = indexMetadata.getSettings().get(IndexModule.INDEX_STORE_TYPE_SETTING.getKey());
                    if (storeType.equals(IndexModule.Type.REMOTE_SNAPSHOT.getSettingsKey())) {
                        indexSearchableSnapshotIds.add(indexMetadata.getSettings().get(IndexSettings.SEARCHABLE_SNAPSHOT_ID_UUID.getKey()));
                    }
                });

                if (snapshotIds.stream().anyMatch(indexSearchableSnapshotIds::contains)) {
                    throw new SnapshotException(
                        repoName,
                        snapshotsToBeNotDeleted.toString(),
                        "Some remote snapshots are supporting searchable snapshot indices and hence can't be deleted! No snapshots were deleted."
                    );
                }
```

Reply:
Thanks for the comment. Since we need to display which all remote searchable snapshots back at least 1 index, simply throwing the exception as soon as we encounter such snapshot won't suffice. We would need to loop over all such remote snapshots and do the check inside that loop.

Reply:
If I understand correctly, I think we can also achieve the same with this block. It saves on the nested loop cost given that number of indices > number of snapshots to be deleted and the relation of index to snapshot is 1:1.

```suggestion
            private void filterSnapshotsBackingAnyIndex(ClusterState currentState, List<SnapshotId> snapshotIds) {
                final Map<String, SnapshotId> uuidToSnapshotId = new HashMap<>();
                final Set<String> snapshotsToBeNotDeleted = new HashSet<>();
                ImmutableOpenMap<String, IndexMetadata> indicesMap = currentState.getMetadata().getIndices();
                snapshotIds.forEach(snapshotId -> uuidToSnapshotId.put(snapshotId.getUUID(), snapshotId));
                
                for (Iterator<IndexMetadata> it = indicesMap.valuesIt(); it.hasNext();) {
                    IndexMetadata indexMetadata = it.next();
                    String storeType = indexMetadata.getSettings().get(IndexModule.INDEX_STORE_TYPE_SETTING.getKey());
                    if (IndexModule.Type.REMOTE_SNAPSHOT.getSettingsKey().equals(storeType)) {
                         String snapshotId = indexMetadata.getSettings().get(IndexSettings.SEARCHABLE_SNAPSHOT_ID_UUID.getKey());
                         if (uuidToSnapshotId.get(snapshotId) != null) {
                             snapshotsToBeNotDeleted.add(uuidToSnapshotId.get(snapshotId).getName());
                         }
                    }
                }
                
                if (!snapshotsToBeNotDeleted.isEmpty()) {
                    throw new SnapshotException(
                        repoName,
                        snapshotsToBeNotDeleted.toString(),
                        "These remote snapshots are backing some indices and hence can't be deleted! No snapshots were deleted."
                    );
                }
            }
```

Reply:
Made the suggested change